engine/src/main/java/io/zeebe/engine/processing/common/EventTriggerBehavior.java

```diff
@@ -0,0 +1,279 @@
+
+            triggerInterruptingEventSubProcess(flowScopeContext, record, eventElementInstanceKey);
+
+          } else {
+            // activate non-interrupting event sub-process
+            publishActivatingEvent(eventElementInstanceKey, record);
+          }
+
+          return eventElementInstanceKey;
+        });
+  }
+
+  private void triggerInterruptingEventSubProcess(
+      final BpmnElementContext flowScopeContext,
+      final ProcessInstanceRecord eventRecord,
+      final long eventElementInstanceKey) {
+
+    unsubscribeFromEvents(flowScopeContext);
+
+    final var noActiveChildInstances = terminateChildInstances(flowScopeContext);
+    if (noActiveChildInstances) {
+      // activate interrupting event sub-process
+      publishActivatingEvent(eventElementInstanceKey, eventRecord);
+
+    } else {
+      // wait until child instances are terminated
+      deferActivatingEvent(
+          flowScopeContext.getElementInstanceKey(), eventElementInstanceKey, eventRecord);
+    }
+
+    elementInstanceState.updateInstance(
+        flowScopeContext.getElementInstanceKey(),
+        flowScopeInstance -> flowScopeInstance.setInterruptingEventKey(eventElementInstanceKey));
+  }
+
+  private boolean terminateChildInstances(final BpmnElementContext flowScopeContext) {
+    final var childInstances =
+        elementInstanceState.getChildren(flowScopeContext.getElementInstanceKey()).stream()
+            .map(
+                childInstance ->
+                    flowScopeContext.copy(
+                        childInstance.getKey(), childInstance.getValue(), childInstance.getState()))
+            .collect(Collectors.toList());
+
+    for (final BpmnElementContext childInstanceContext : childInstances) {
+
+      if (ProcessInstanceLifecycle.canTerminate(childInstanceContext.getIntent())) {
+        if (!MigratedStreamProcessors.isMigrated(childInstanceContext.getBpmnElementType())) {
+          transitionToTerminating(childInstanceContext);
+        } else {
+          commandWriter.appendFollowUpCommand(
+              childInstanceContext.getElementInstanceKey(),
+              ProcessInstanceIntent.TERMINATE_ELEMENT,
+              childInstanceContext.getRecordValue());
+        }
+
+      } else if (childInstanceContext.getIntent() == ProcessInstanceIntent.ELEMENT_COMPLETED) {
+        // clean up the state because the completed event will not be processed
+        eventScopeInstanceState.deleteInstance(childInstanceContext.getElementInstanceKey());
+        elementInstanceState.removeInstance(childInstanceContext.getElementInstanceKey());
+      }
+    }
+
+    final var elementInstance =
+        elementInstanceState.getInstance(flowScopeContext.getElementInstanceKey());
+    final var activeChildInstances = elementInstance.getNumberOfActiveElementInstances();
+
+    return activeChildInstances == 0;
+  }
+
+  public BpmnElementContext transitionToTerminating(final BpmnElementContext context) {
```

Comment:
As I mentioned in one of our discussions and also discussed with @saig0 this is an intermediate solution and in any case it will be either removed or changed in some way, since we want to merge or refactor several behaviors. I also dont like it, but it was the easiest to make progress and finish this stuff. Priority is for me now migrating the processors.

Reply:
I wasn't sure if this was something that'll be removed once everything is migrated - if so, please add a TODO as we do for other similar things, so we can easily find and remove them later.